### PR TITLE
[SPARK-26765][SQL] Avro: Validate input and output schema

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

The API `supportDataType` in `FileFormat` helps to validate the output/input schema before exection starts. So that we can avoid some invalid data source IO, and users can see clean error messages.

This PR is to override the validation API in Avro data source. 
Also, as per the spec of Avro(https://avro.apache.org/docs/1.8.2/spec.html), `NullType` is supported. This PR fixes the handling of `NullType`.

## How was this patch tested?

Unit test

### CODE REVIEW

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 1]
```diff
@@ -243,6 +243,23 @@ private[avro] class AvroFileFormat extends FileFormat
...
+    case ArrayType(elementType, _) => supportDataType(elementType)
+
+    case MapType(keyType, valueType, _) =>
+      supportDataType(keyType) && supportDataType(valueType)
+
+    case udt: UserDefinedType[_] => supportDataType(udt.sqlType)
+
+    case _: NullType => true
+
+    case _ => false
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: which type avro doesn't support?

Comment (gengliangwang) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: E.g Interval data type. The supported types are the same as JSON.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
+1, late LGTM.

**Comment by HyukjinKwon** [comment]
Ah, later LGTM :D.

**Pull request closed without merging**